### PR TITLE
5.2: Key rotation

### PR DESCRIPTION
**Pull request by nadinequinr3**
Preview: https://nadine.preview.docs.r3.com/en/platform/corda/5.2/deploying-operating/keys.html

### CODE REVIEW

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  corda52:
+    identifier: corda52-cluster-wrapping-keys
+    parent: corda52-cluster
+    weight: 3025
+---
+
+# Managing Wrapping Keys
+
+As described in the [Architecture for Cluster Administrators]({{< relref "../../key-concepts/cluster-admin/_index.md#key-management" >}}) section, Corda keys are stored in the cluster and virtual node `Crypto` databases and encrypted at rest with wrapping keys.
+These managed wrapping keys are in turn wrapped by a master wrapping key.
```
Comment (simon-johnson-r3) [inline content/en/platform/corda/5.2/deploying-operating/keys/_index.md:15]: I don't think "managed wrapping keys" is a term end users will understand without more explanation.

How about:
These wrapping keys are created and managed automatically by Corda, hence we refer to them as managed wrapping keys. These managed wrapping keys are in turn encrypted with a master wrapping key and also stored in the cluster and virtual node `Crypto` databases.

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+This section describes how Corda stores the master wrapping key and how to rotate master and managed wrapping keys. It contains the following:
+
+{{< toc >}}
+
+## Configuring the Master Wrapping Key
+
+The master wrapping key, or the information required to generate this key, must be stored and managed outside Corda.
+You can achieve this in one of the following ways:
+
+* Pass a passphrase and salt, to generate the master key, into the crypto worker processes. For more information, see [Default Secrets Service]({{< relref "../../deploying-operating/deployment/deploying/_index.md#default-secrets-service" >}}).
```
Comment (simon-johnson-r3) [inline content/en/platform/corda/5.2/deploying-operating/keys/_index.md:25]: `crypto worker processes` is an implementation detail customers don't really need to understand at this level. Also if you you use this method you aren't really storing outside Corda as the previous line suggests.

Suggestion:
The master wrapping key must be stored and managed outside Corda when running a production deployment. For this an external key management system is used to retrieve the master key when required. For more information, see [External Secrets Service]({{< relref "../../deploying-operating/deployment/deploying/_index.md#external-secrets-service" >}}).

For testing, development or other non production uses, a passphrase and salt can be provided to generate the master key, and the master key will then be stored by Corda. For more information, see [Default Secrets Service]({{< relref "../../deploying-operating/deployment/deploying/_index.md#default-secrets-service" >}}).

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+* {{< enterprise-icon noMargin="true" >}} Store and manage the master in an external key management system for Corda to retrieve when required. For more information, see [External Secrets Service]({{< relref "../../deploying-operating/deployment/deploying/_index.md#external-secrets-service" >}}).
+
+Corda stores the master wrapping key settings in the `corda.crypto` configuration section.
```
Comment (simon-johnson-r3) [inline content/en/platform/corda/5.2/deploying-operating/keys/_index.md:28]: Given the above comment, perhaps change to "In both cases, Corda stores..."

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+                "encryptedSecret": "cvNVJ4D7fNGm59Uh3Tlhi0W3pToDxl7vykUdZaxEuEPGUMdmu7G+xTEfl0CcC2R9aWl6FgrzmX5x+qCGNElUi/PBA40EGjEU"
+              }
+            },
+          },
+          {
+            "alias": "master2",
+            "passphrase":{
+              "configSecret":
+              {
+                "encryptedSecret": "ZAmGcppwqUNsViSE06b801DVARbD9wT7SZnYP03WKZFOnC9KTsII04pWWq5Je2CUZdP+xIsHFUO3RkZVwhwnyQhWoxcW6JEo"
```
Comment (simon-johnson-r3) [inline content/en/platform/corda/5.2/deploying-operating/keys/_index.md:103]: I think we should describe how a user would generate this. Also perhaps we need to point out this is the config for a Corda internal master key, and also provide an example for an externally managed one.

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+* [Master Wrapping Key](#rotating-master-wrapping-key)
+* [Managed Wrapping Keys](#rotating-managed-wrapping-keys)
+
+### Rotating Master Wrapping Key
+
+Because the master key is not managed by Corda, you must first change the default key in the source system.
+Corda must then re-encrypt all of the keys that were wrapped with the old master key.
+To rotate the master wrapping key, do the following:
+
+1. Add the new key to the source environment. For more information, see [Configuring the Master Wrapping Key](#configuring-the-master-wrapping-key).
```
Comment (simon-johnson-r3) [inline content/en/platform/corda/5.2/deploying-operating/keys/_index.md:46]: Add the new **master** key...

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+2. Add the new key to the `corda.crypto` configuration section, setting it as the default key. The [Setting Configuration Fields Dynamically]({{< relref "../config/dynamic.md#setting-configuration-fields-dynamically" >}}) section describes how to update configuration values. For example, to add a new master key with the alias `master2`:
```
Comment (simon-johnson-r3) [inline content/en/platform/corda/5.2/deploying-operating/keys/_index.md:47]: Perhaps add "...  from now on all new managed wrapping keys will be encrypted with this new master key"

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        "major": 1,
+        "minor": 0
+      },
+      "section": "corda.crypto",
+      "version": 1
+   })
+   ```
+   {{% /tab %}}
+   {{< /tabs >}}
+3. Rotate the key to the new default key using the POST method of the [/api/v5_2/wrappingkey/rotation/{tenantid} endpoint](../../reference/rest-api/openapi.html#tag/Key-Rotation-API/operation/post_wrappingkey_rotation__tenantid_):
```
Comment (simon-johnson-r3) [inline content/en/platform/corda/5.2/deploying-operating/keys/_index.md:132]: Rotate the old master key to the new default master key. After this step all managed wrapping keys will be encrypted with the new master key.

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+   curl -k -u $REST_API_USER:$REST_API_PASSWORD -X GET "$REST_API_URL/wrappingkey/rotation/master"
+   ```
+   {{% /tab %}}
+   {{% tab name="PowerShell" %}}
+   ```shell
+  Invoke-RestMethod -SkipCertificateCheck -Headers @{Authorization=("Basic {0}" -f ${REST_API_USER}:${REST_API_PASSWORD})} -Method GET -Uri "$REST_API_URL/wrappingkey/rotation/master" 
+   ```
+   {{% /tab %}}
+   {{< /tabs >}}
+
```
Comment (simon-johnson-r3) [inline content/en/platform/corda/5.2/deploying-operating/keys/_index.md:158]: 4. At this point all new wrapping keys are created using the new master key and all existing wrapping keys are rewrapped with the new master key. You can now remove the reference to the old master wrapping key from Corda by updating the config. You can also remove the key from any external key management system.

File: content/en/platform/corda/5.2/key-concepts/cluster-admin/_index.md [hunk 9]
```diff
@@ -96,10 +96,9 @@ All keys are stored in the Crypto databases (cluster and virtual nodes) and they
...
+The diagram illustrates that key wrapping is hierarchical. The master wrapping key protects other wrapping keys, such as the virtual node related keys, which in turn protect the private keys used by Corda. The master wrapping key, or the information required to generate this key, must be stored and managed outside Corda. This ensures that it is not possible to decrypt the keys stored in the database using other information stored in the same database. By default, the Corda default secrets lookup service generates the master key using a passphrase and salt set for the crypto worker processes. {{< enterprise-icon noMargin="true" >}}Alternatively, Corda Enterprise can retrieve the master key from an external key management system that Corda retrieves when required.
```
Comment (simon-johnson-r3) [inline content/en/platform/corda/5.2/key-concepts/cluster-admin/_index.md:99]: "The master wrapping key, or the information required to generate this key..." see comment above, I'm not sure this is quite explaining the situation. Production systems would manage the key externally, non-production can use the built in default secrets service.

**Review by simon-johnson-r3 — CHANGES_REQUESTED** [review]

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md (hunk 4 above)
Comment (nadinequinr3) [reply content/en/platform/corda/5.2/deploying-operating/keys/_index.md:101]: @simon-johnson-r3 Could you provide me with sample values? `vaultPath` and `vaultKey` I think?

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+        "major": 1,
+        "minor": 0
+      },
+      "section": "corda.crypto",
+      "version": 1
+   })
+   ```
+   {{% /tab %}}
+   {{< /tabs >}}
+   Corda wraps all new managed wrapping keys with this new master wrapping key.
```
Comment (Sakpal) [inline content/en/platform/corda/5.2/deploying-operating/keys/_index.md:130]: Minor: We have the following statement after step 3 `Corda rewraps all existing managed wrapping keys with the new master wrapping key.`

Shall we skip the statement in line 130 as it seems like a repetition?

Comment (nadinequinr3) [reply content/en/platform/corda/5.2/deploying-operating/keys/_index.md:130]: My understanding is that after step 2, all **new** managed wrapping keys will be rewrapped with the new master key. After step 3,  all **existing** managed wrapping keys are wrapped with the new master wrapping key.

I've modified it slightly to make this clearer.

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    "{\"caching\ :\"expireAfterAccessMins\":{\"default\":60},\"maximumSize\" :{\"default\":10000}},\"hsm\": \
+    {\"defaultWrappingKey\":\"master2\",\"retrying\": {\"attemptTimeoutMills\":20000, \"maxAttempts\":3}, \"wrappingKeys\": \
+    [{\"alias\":\"master1\", \"passphrase\": {\"configSecret\": {\"encryptedSecret\" :\"ZAmGcppwqUNsViSE06b801DVARbD9wT7SZnYP03WKZFOnC9KTsII04pWWq5Je2CUZdP+xIsHFUO3RkZVwhwnyQhWoxcW6JEo\"}},\"salt\": \
+    {\"configSecret\":{\"encryptedSecret\":\"cvNVJ4D7fNGm59Uh3Tlhi0W3pToDxl7vykUdZaxEuEPGUMdmu7G+xTEfl0CcC2R9aWl6FgrzmX5x+qCGNElUi/PBA40EGjEU\"}}}, \
+    {\"alias\":\"master2\",\"passphrase\":{\"configSecret\": {\"encryptedSecret\":\"gzXRjE4AoZSW8GQhcqbw0shQfwWGy5O8alNDqXaqTUzyUnLBVUXLwhv+uAEZzecQixiaEaNtfOvuftYW9NjwwRyf6m4KZIag\"}},\"salt\": \
+    {\"configSecret\":{\"encryptedSecret\":\"flvAnf5G/aRmDIaahXHjftEpc0odufhzZsuiipxtQTFV+pccB+RHnIqle26jjDlDUBfn3HPbVezKhpRHL/NP5M+cX/rU15DE\"}}}]}, \
+    \"retrying\":{\"maxAttempts\": {\"default\": 3},\"waitBetweenMills\":{\"default\":[200]}}}}","schemaVersion": {"major":1, "minor":0}}' \
+    "$REST_API_URL/config"
```
Comment (anien) [inline content/en/platform/corda/5.2/deploying-operating/keys/_index.md:58]: I can't get this command working.
But this one works fine (with different secrets):
```kotlin
curl -k -u admin:admin  -X 'PUT' \
  'https://localhost:8888/api/v5_2/config' \
  -d '{
  "section": "corda.crypto",
  "config": "{\"caching\":{\"expireAfterAccessMins\":{\"default\":60},\"maximumSize\":{\"default\":10000}},\"hsm\":{\"defaultWrappingKey\":\"master2\",\"retrying\":{\"attemptTimeoutMills\":20000,\"maxAttempts\":3},\"wrappingKeys\":[{\"alias\":\"master1\",\"passphrase\":{\"configSecret\":{\"encryptedSecret\":\"ZAmGcppwqUNsViSE06b801DVARbD9wT7SZnYP03WKZFOnC9KTsII04pWWq5Je2CUZdP+xIsHFUO3RkZVwhwnyQhWoxcW6JEo\"}},\"salt\":{\"configSecret\":{\"encryptedSecret\":\"cvNVJ4D7fNGm59Uh3Tlhi0W3pToDxl7vykUdZaxEuEPGUMdmu7G+xTEfl0CcC2R9aWl6FgrzmX5x+qCGNElUi/PBA40EGjEU\"}}},{\"alias\":\"master2\",\"passphrase\":{\"configSecret\":{\"encryptedSecret\":\"gzXRjE4AoZSW8GQhcqbw0shQfwWGy5O8alNDqXaqTUzyUnLBVUXLwhv+uAEZzecQixiaEaNtfOvuftYW9NjwwRyf6m4KZIag\"}},\"salt\":{\"configSecret\":{\"encryptedSecret\":\"flvAnf5G/aRmDIaahXHjftEpc0odufhzZsuiipxtQTFV+pccB+RHnIqle26jjDlDUBfn3HPbVezKhpRHL/NP5M+cX/rU15DE\"}}}]},\"retrying\":{\"maxAttempts\":{\"default\":3},\"waitBetweenMills\":{\"default\":[200]}}}",
  "schemaVersion": {
    "major": 1,
    "minor": 0
  },
  "version": 0
}'
```

**Review by anien — COMMENTED** [review]
Just one minor comment, otherwise it looks good to me!

**Review by anien — APPROVED** [review]
LGTM!

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md (hunk 10 above)
Comment (nadinequinr3) [reply content/en/platform/corda/5.2/deploying-operating/keys/_index.md]: Second example added in https://github.com/corda/corda-docs-portal/pull/2047/commits/42466aad242375a9636bdeeac210ada7f6cbce91

File: content/en/platform/corda/5.2/deploying-operating/keys/_index.md (hunk 4 above)
Comment (nadinequinr3) [reply content/en/platform/corda/5.2/deploying-operating/keys/_index.md:101]: Example added in https://github.com/corda/corda-docs-portal/pull/2047/commits/42466aad242375a9636bdeeac210ada7f6cbce91

File: content/en/platform/corda/5.2/key-concepts/cluster-admin/_index.md [hunk 12]
```diff
@@ -96,10 +96,9 @@ All keys are stored in the Crypto databases (cluster and virtual nodes) and they
...
+The diagram illustrates that key wrapping is hierarchical. The master wrapping key protects other wrapping keys, such as the virtual node related keys, which in turn protect the private keys used by Corda. The master wrapping key, or the information required to generate this key, must be stored and managed outside the standard Corda databases. This ensures that it is not possible to decrypt the keys stored in the database using other information stored in the same database. In testing, development, and or other non-production environments, the Corda default secrets lookup service can generate the master key using a passphrase and salt set for the crypto worker processes. In production environments, Corda Enterprise retrieves the master key from an external key management system when required.
```
Comment (AnnaK-R3) [inline content/en/platform/corda/5.2/key-concepts/cluster-admin/_index.md:99]:
```suggestion
The diagram illustrates that key wrapping is hierarchical. The master wrapping key protects other wrapping keys, such as the virtual node related keys, which in turn protect the private keys used by Corda. The master wrapping key, or the information required to generate this key, must be stored and managed outside the standard Corda databases. This ensures that it is not possible to decrypt the keys stored in the database using other information stored in the same database. In testing, development, and other non-production environments, the Corda default secrets lookup service can generate the master key using a passphrase and salt set for the crypto worker processes. In production environments, Corda Enterprise retrieves the master key from an external key management system when required.
```

**Review by AnnaK-R3 — APPROVED** [review]

**Pull request merged**